lib/object-path-immutable-transaction.js

```diff
@@ -0,0 +1,98 @@
+'use strict';
+const _assign = require('lodash.assign');
+const objectPath = require('object-path');
+const _isArray = require('lodash.isarray');
+const _isPlainObject = require('lodash.isplainobject');
+
+function ensureCloned(path, source, dest) {
+	if (!_isArray(path)) {
+		path = ('' + path).split('.');
+	}
+
+	// ensure all ancestor path segments are cloned
+	for (let i = 0; i < path.length; i++) {
+		if (dest[path[i]] === source[path[i]]) {
+			if (_isArray(source[path[i]])) {
+				dest[path[i]] = source[path[i]].slice();
+			} else {
+				dest[path[i]] = _assign({}, source[path[i]]);
+			}
+		}
+
+		source = source[path[i]];
+		dest = dest[path[i]];
+	}
+}
+
+function createTransaction(source) {
+	if (!_isPlainObject(source)) {
+		throw new Error('Source must be a plain object. Got ' + source);
+	}
+
+	const result = _assign({}, source);
```

Comment:
If I understand the use case for this correctly, someone would call:

```
objectPathImmutable. createTransaction(o)
.set('foo.bar.applesauce', 'chutney');
.set('foo.bar.jam', 'marmelade')
.commit();
```

Doesn't `ensureCloned` clone each object in source every time it is called? It gets called for each call to `set`. Wouldn't `foo` and `bar` both be cloned 2 times in that example? That is the same number as the example given in the issue.

Is a goal of this PR to minimize how many times the source object and its children are cloned?

What if I call:

```
let o = { foo: { bar: {} } };
objectPathImmutable. createTransaction(o)
.set('foo.bar.applesauce', 'chutney')
.set('foo.bar.jam', 'marmelade')
.delete('foo')
.commit();
```

Wouldn't this transaction have cloned `source`, `source.foo`, and `source.foo.bar` before returning `{}`?

The referenced issue #7 discusses being lazy as part of the solution, but this code does not seem to be lazy.

Would something like this benefit from lodash's shortcut fusion applied to the chain method?
https://lodash.com/docs#chain